File: Language/C#/Book/tmp/LINQ实战/LinqInAction/Chapter06to08.CSharp/Common/SampleClasses/Ch6/Subject.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Data.Linq;
using System.Data.Linq.Mapping;
using System.Text;
using System.Data.SqlClient;

namespace LinqInAction.Chapter06to08.Common.SampleClasses.Ch6.Unmapped
{
  /// <summary>
  /// Subject class definition from listing 6.3
  /// </summary>
  public class Subject
  {
    public Guid SubjectId { get; set; }
    public String Description { get; set; }
    public String Name { get; set; }

    public static IEnumerable<Subject> GetSubjects()
    {
      List<Subject> subjects = new List<Subject>();
      using (SqlConnection connection = new SqlConnection(LinqInAction.Chapter06to08.Common.Properties.Settings.Default.liaConnectionString))
      {
        connection.Open();
        SqlCommand command = connection.CreateCommand();
        command.CommandText = "SELECT ID, Name, Description FROM dbo.Subject";
        using (SqlDataReader reader = command.ExecuteReader())
        {
          while (reader.Read())
          {
            Subject newSubject = new Subject();
            if (!reader.IsDBNull(0)) { newSubject.SubjectId = reader.GetGuid(0); }
            if (!reader.IsDBNull(1)) { newSubject.Description = reader.GetString(1); }
            if (!reader.IsDBNull(2)) { newSubject.Name = reader.GetString(2); }
            subjects.Add(newSubject);
          }
        }
      }
      return subjects;
    }
  }
}
namespace LinqInAction.Chapter06to08.Common.SampleClasses.Ch6
{
  /// <summary>
  /// Subject class definition from listing 6.3
  /// </summary>
  [Table]
  public class Subject
  {
    public Subject()
    {
      this._Books = new EntitySet<Book>(new Action<Book>(this.attach_Books), new Action<Book>(this.detach_Books));
    }

    [Column(IsPrimaryKey = true, Name = "ID")]
    public Guid SubjectId { get; set; }
    [Column]
    public String Description { get; set; }
    [Column]
    public String Name { get; set; }

    public static IEnumerable<Subject> GetSubjects()
    {
      DataContext context = new DataContext(LinqInAction.Chapter06to08.Common.Properties.Settings.Default.liaConnectionString);
      return context.GetTable<Subject>();
    }


    private EntitySet<Book> _Books;

    [Association(Name = "Subject_Book", Storage = "_Books", OtherKey = "SubjectId")]
    public EntitySet<Book> Books
    {
      get
      {
        return this._Books;
      }
      set
      {
        this._Books.Assign(value);
      }
    }

    private void attach_Books(Book entity)
    {
      // entity.Subject1 = this;
    }

    private void detach_Books(Book entity)
    {
      // entity.Subject1 = null;
    }
  }
}

File: Language/C#/Book/LINQ实战/第06章 LINQ to SQL入门/Chapter06to08.VB/Common/SampleClasses/Ch6/Subject.vb
Imports System.Data.Linq.Mapping
Imports System.Data.Linq
Imports System.Data.SqlClient

Namespace SampleClasses.Ch6.Unmapped
    ''' <summary>
    ''' Subject class definition from listing 6.3
    ''' </summary>
    Public Class Subject

        Private _SubjectId As Guid
        Public Property SubjectId() As Guid
            Get
                Return _SubjectId
            End Get
            Set(ByVal value As Guid)
                _SubjectId = value
            End Set
        End Property


        Private _Description As String
        Public Property Description() As String
            Get
                Return _Description
            End Get
            Set(ByVal value As String)
                _Description = value
            End Set
        End Property


        Private _Name As String
        Public Property Name() As String
            Get
                Return _Name
            End Get
            Set(ByVal value As String)
                _Name = value
            End Set
        End Property

        Public Shared Function GetSubjects() As IEnumerable(Of Subject)
            Dim subjects As New List(Of Subject)
            Using connection As New System.Data.SqlClient.SqlConnection(My.Settings.liaConnectionString)
                connection.Open()
                Using command As SqlCommand = connection.CreateCommand()
                    command.CommandText = "SELECT ID, Name, Description FROM dbo.Subject"
                    Using reader As SqlDataReader = command.ExecuteReader()
                        While (reader.Read())
                            Dim newSubject As New Subject()
                            If Not reader.IsDBNull(0) Then newSubject.SubjectId = reader.GetGuid(0)
                            If Not reader.IsDBNull(1) Then newSubject.Description = reader.GetString(1)
                            If Not reader.IsDBNull(2) Then newSubject.Name = reader.GetString(2)
                            subjects.Add(newSubject)
                        End While
                    End Using
                End Using
            End Using
            Return subjects
        End Function
    End Class
End Namespace

Namespace SampleClasses.Ch6
    ''' <summary>
    ''' Subject class definition from listing 6.3
    ''' </summary>
    <Table()> _
    Public Class Subject
        Public Sub New()
            Me._Books = New EntitySet(Of Book)(AddressOf Me.attach_Books, AddressOf Me.detach_Books)
        End Sub

        Private _SubjectId As Guid
        <Column(IsPrimaryKey:=True, Name:="ID")> _
        Public Property SubjectId() As Guid
            Get
                Return _SubjectId
            End Get
            Set(ByVal value As Guid)
                _SubjectId = value
            End Set
        End Property


        Private _Description As String
        <Column()> _
        Public Property Description() As String
            Get
                Return _Description
            End Get
            Set(ByVal value As String)
                _Description = value
            End Set
        End Property


        Private _Name As String
        <Column()> _
        Public Property Name() As String
            Get
                Return _Name
            End Get
            Set(ByVal value As String)
                _Name = value
            End Set
        End Property


        Public Shared Function GetSubjects() As IEnumerable(Of Subject)
            Dim context As New DataContext(My.Settings.liaConnectionString)
            Return context.GetTable(Of Subject)()
        End Function

        Private _Books As EntitySet(Of Book)

        <Association(Name:="Subject_Book", Storage:="_Books", OtherKey:="SubjectId")> _
        Public Property Books() As EntitySet(Of Book)
            Get
                Return _Books
            End Get
            Set(ByVal value As EntitySet(Of Book))
                _Books.Assign(value)
            End Set
        End Property

        Private Sub attach_Books(ByVal entity As Book)
            '        entity.Subject1 = Me
        End Sub

        Private Sub detach_Books(ByVal entity As Book)
            ' entity.Subject1 = nothing
        End Sub
    End Class
End Namespace